pkg/cable/vxlan/vxlan.go

```diff
@@ -0,0 +1,538 @@
+	ipSlice := strings.Split(ipAddr, ".")
+	if len(ipSlice) < 4 {
+		return nil, fmt.Errorf("invalid ipAddr [%s]", ipAddr)
+	}
+
+	ipSlice[0] = strconv.Itoa(VxlanVTepNetworkPrefix)
+	vxlanIP := net.ParseIP(strings.Join(ipSlice, "."))
+
+	return vxlanIP, nil
+}
+
+func (v *vxlan) addIPRule() error {
+	if v.vxlanIface != nil {
+		rule := netlink.NewRule()
+		rule.Table = TableID
+		rule.Priority = TableID
+
+		err := netlink.RuleAdd(rule)
+		if err != nil && !os.IsExist(err) {
+			return fmt.Errorf("failed to add ip rule %s: %v", rule, err)
+		}
+	}
+
+	return nil
+}
+
+func (v *vxlan) ConnectToEndpoint(endpointInfo *natdiscovery.NATEndpointInfo) (string, error) {
+	remoteEndpoint := endpointInfo.Endpoint
+	if v.localEndpoint.Spec.ClusterID == remoteEndpoint.Spec.ClusterID {
+		klog.V(log.DEBUG).Infof("Will not connect to self")
+		return "", nil
+	}
+
+	remoteIP := net.ParseIP(endpointInfo.UseIP)
+	if remoteIP == nil {
+		return "", fmt.Errorf("failed to parse remote IP %s", endpointInfo.UseIP)
+	}
+
+	allowedIPs := parseSubnets(remoteEndpoint.Spec.Subnets)
+
+	klog.V(log.DEBUG).Infof("Connecting cluster %s endpoint %s",
+		remoteEndpoint.Spec.ClusterID, remoteIP)
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+
+	cable.RecordConnection(CableDriverName, &v.localEndpoint.Spec, &remoteEndpoint.Spec, string(v1.Connected), true)
+
+	remoteVtepIP, err := v.getVxlanVtepIPAddress(endpointInfo.UseIP)
+
+	if err != nil {
+		return endpointInfo.UseIP, fmt.Errorf("failed to derive the vxlan vtepIP for %s, %v", endpointInfo.UseIP, err)
+	}
+
+	err = v.vxlanIface.AddFDB(remoteIP, "00:00:00:00:00:00")
+
+	if err != nil {
+		return endpointInfo.UseIP, fmt.Errorf("failed to add remoteIP %q to the forwarding database", remoteIP)
+	}
+
+	err = v.vxlanIface.AddRoute(allowedIPs, remoteVtepIP, v.vxlanIface.vtepIP)
+
+	if err != nil {
+		return endpointInfo.UseIP, fmt.Errorf("failed to add route for the CIDR %q with remoteVtepIP %q and vxlanInterfaceIP %q",
+			allowedIPs, remoteVtepIP, v.vxlanIface.vtepIP)
+	}
+
+	v.connections = append(v.connections, v1.Connection{Endpoint: remoteEndpoint.Spec, Status: v1.Connected,
+		UsingIP: endpointInfo.UseIP, UsingNAT: endpointInfo.UseNAT})
+
+	klog.V(log.DEBUG).Infof("Done adding endpoint for cluster %s", remoteEndpoint.Spec.ClusterID)
+
+	return endpointInfo.UseIP, nil
+}
+
+func (v *vxlan) DisconnectFromEndpoint(remoteEndpoint types.SubmarinerEndpoint) error {
+	klog.V(log.DEBUG).Infof("Removing endpoint %#v", remoteEndpoint)
+
+	if v.localEndpoint.Spec.ClusterID == remoteEndpoint.Spec.ClusterID {
+		klog.V(log.DEBUG).Infof("Will not disconnect self")
+		return nil
+	}
+
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+
+	var ip string
+
+	for _, connection := range v.connections {
+		if connection.Endpoint.CableName == remoteEndpoint.Spec.CableName {
+			ip = connection.UsingIP
+		}
+	}
+
+	if ip == "" {
+		return fmt.Errorf("failed to get remote IP %s", remoteEndpoint.Spec.CableName)
```

Comment:
If we return an error, this will keep retrying, but the IP will not be found in the future (may be
if the endpoint is re-added, but at that time we really don't want to remove it anymore)

In this case I believe we should log the error and return nil.
```suggestion
		klog.Errorf("failed to get remote IP %s", remoteEndpoint.Spec.CableName)
		return nil
```

Reply:
yeah I thought the same thing but callers don't specifically re-queue. It's fine either way although the caller does provide more context in the message. So if we want to log here then we could say "_Cannot disconnect remote endpoint 
 %q - no prior connection entry found_".

Reply:
Done.